spirv_hlsl.cpp

```diff
@@ -6693,3 +6718,15 @@ bool CompilerHLSL::builtin_translates_to_nonarray(spv::BuiltIn builtin) const
 {
 	return (builtin == BuiltInSampleMask);
 }
+
+bool CompilerHLSL::is_user_type_structured(uint32_t id) const
+{
+	if (hlsl_options.preserve_structured_buffers)
+	{
+		// Compare left hand side of string only as these user types can contain more meta data such as their subtypes,
+		// e.g. "structuredbuffer:int"
+		const std::string &user_type = get_decoration_string(id, DecorationUserTypeGOOGLE);
+		return user_type.compare(0, 16, "structuredbuffer") == 0 || user_type.compare(0, 18, "rwstructuredbuffer") == 0;
```

Comment:
Is there any risk of DXC optimizing away members of the struct? spirv-opt has been known to do this lately where the struct may be laid out as:

```
struct Foo
{
    [offset = 0] int a;
    [offset = 12] int d;
};
```

where the original was something like

```
struct Foo
{
   int a, b, c, d;
};
```
but b and c were unused. We'd need to make sure somehow that the layout works out.

Reply:
There is indeed a SPIRV-Tools pass that removes unused struct members, but it's only part of `-Os` optimization when you optimize for size. In our fork of DXC, we added this particular pass to the performance optimization, i.e. `-O2` etc. but we only apply it on the struct for the special case of `$Globals` cbuffer. If someone applies that optimization pass on structs used for structured buffers they would not only have to retain the offsets but also have potentially a last dummy member or otherwise the struct size would be incorrect if the last element is removed. I assume that optimization would also cause problems in GLSL.

Reply:
Inner structs don't support layout(offset) in GL unfortunately, so it is a problem there as well, but I haven't seen that reported as an issue yet. I'd rather not have to support all of that (MSL buffer packing is a nightmare) just to handle this scenario. If your use case can avoid making SPIRV-Cross' life difficult, that would be preferable.

We'd probably need to just run the "does this struct conform to HLSL structured buffer rules" to sanity check that nothing weird is going to happen. Likely a new buffer layout type ala HLSLCbuffer would have to be added for that.